File: node-sdk/src/metadata.ts
/**
 * Secret metadata parsing.
 */

import { readFileSync } from "node:fs";
import yaml from "js-yaml";

/** Plaintext metadata for a single secret. */
export interface SecretMetadata {
  /** Secret name identifier. */
  name: string;
  /** Group this secret belongs to. */
  group: string;
  /** When the secret was first created. */
  createdAt: Date;
  /** When the secret was last rotated. */
  rotatedAt: Date;
  /** Optional expiration date. */
  expiresAt?: Date;
  /** List of agent names authorized to access this secret. */
  authorizedAgents: string[];
}

/**
 * Parse a datetime value from YAML, which may be a Date object or ISO string.
 */
function parseDate(val: unknown): Date {
  if (val instanceof Date) {
    return val;
  }
  if (typeof val === "string") {
    const d = new Date(val);
    if (!isNaN(d.getTime())) {
      return d;
    }
  }
  return new Date(0);
}

/**
 * Parse a .meta YAML file into a SecretMetadata object.
 *
 * @param filePath - Absolute path to the .meta file.
 * @returns Parsed SecretMetadata.
 */
export function parseMetadataFile(filePath: string): SecretMetadata {
  const content = readFileSync(filePath, "utf-8");
  return parseMetadata(content);
}

/**
 * Parse metadata from a YAML string.
 *
 * @param content - YAML string content.
 * @returns Parsed SecretMetadata.
 */
export function parseMetadata(content: string): SecretMetadata {
  const data = (yaml.load(content) as Record<string, unknown>) ?? {};

  return {
    name: typeof data.name === "string" ? data.name : "",
    group: typeof data.group === "string" ? data.group : "",
    createdAt: parseDate(data.created),
    rotatedAt: parseDate(data.rotated),
    expiresAt: data.expires ? parseDate(data.expires) : undefined,
    authorizedAgents: Array.isArray(data.authorized_agents)
      ? (data.authorized_agents as string[])
      : [],
  };
}

File: python-sdk/src/agent_vault/metadata.py
"""Secret metadata parsing."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional

import yaml


@dataclass
class SecretMetadata:
    """Plaintext metadata for a single secret."""

    name: str
    group: str
    created: datetime
    rotated: datetime
    expires: Optional[datetime]
    authorized_agents: list[str]

    @classmethod
    def load(cls, path: Path) -> "SecretMetadata":
        """Load metadata from a .meta YAML file."""
        with open(path, "r") as f:
            data = yaml.safe_load(f) or {}

        return cls(
            name=data.get("name", ""),
            group=data.get("group", ""),
            created=_parse_dt(data.get("created")),
            rotated=_parse_dt(data.get("rotated")),
            expires=_parse_dt(data.get("expires")) if data.get("expires") else None,
            authorized_agents=list(data.get("authorized_agents", [])),
        )


def _parse_dt(val) -> datetime:
    """Parse a datetime from YAML (may be string or datetime)."""
    if isinstance(val, datetime):
        return val
    if isinstance(val, str):
        return datetime.fromisoformat(val)
    return datetime.min
